core/src/main/java/io/undertow/conduits/InflatingStreamSourceConduit.java

```diff
@@ -211,10 +211,14 @@ public int read(ByteBuffer dst) throws IOException {
     }
 
     protected void initializeInflater(ByteBuffer buf) throws IOException {
-        if(isZlibHeaderPresent(buf)) {
-            this.activePooledObject = this.objectPoolWrapping.allocate();
-        } else {
-            this.activePooledObject = this.objectPoolNonWrapping.allocate();
+        // ensure the activePooledObject is set only once until done() is called
+        // we don't want to reset the inflater state mid-stream
+        if (activePooledObject == null) {
+            if (isZlibHeaderPresent(buf)) {
+                this.activePooledObject = this.objectPoolWrapping.allocate();
+            } else {
+                this.activePooledObject = this.objectPoolNonWrapping.allocate();
+            }
```

Comment:
This should likely use some kind of lock or it could still be reset.

Reply:
AFAIR read will happen sequentially in IO thread, so there is no need for that? @fl4via ^^ ?

Reply:
@baranowb is right. We are not supposed to have multiple threads invoking the same conduit at the same time, except for a concurrent close call

Reply:
Sorry, I should probably not have commented that :) I just saw the variable was volatile and assumed it could be accessed by multiple threads at once. If I'm wrong, then nothing is needed at all :)

My concern was the read for the `null` check was it could be `null` then there is a race to set it, but only if multiple threads are invoking the method.